[SPARK-51554][SQL] Add the time_trunc() function for TIME datatype

the-sakthi opened the pull request:
### What changes were proposed in this pull request?
- Added a new built-in function `time_trunc(unit, expr)` that returns a `TIME` value truncated to the specified unit.
- Allowed input for expr to be either a `TIME` type ~~(or a string that can be cast to TIME.)~~
  - Recent addition of the AnyTimeType which is used here, doesn't allow cast from string. Hence this behavior change. However simply allowing anytimetype to be casted from a valid string, this change would automatically be able handle valid input time strings as well.
- Supported truncation units are `HOUR`, `MINUTE`, `SECOND`, `MILLISECOND`, and `MICROSECOND`.
- Handles both foldable and non-foldable inputs


### Why are the changes needed?
- Spark currently lacks a built-in function for truncating TIME values in a similar way as `truncTimestamp`.

### Does this PR introduce _any_ user-facing change?
Yes. A new built-in function `time_trunc` is added. Users can call the function to truncate TIME values to one of the above mentioned supported units.


### How was this patch tested?
By running newly added UTs:
```bash
$ build/sbt "test:testOnly *TimeExpressionsSuite.scala"
```

By manual tests:

```bash
# Happy test cases
scala> spark.sql("SELECT time_trunc('HOUR', TIME'09:32:05.123456');").show()
+---------------------------------+
|time_trunc(HOUR, 09:32:05.123456)|
+---------------------------------+
|                         09:00:00                 |
+---------------------------------+

scala> spark.sql("SELECT time_trunc('MINUTE', TIME'09:32:05.123456');").show()
+------------------------------------------+
|time_trunc(MINUTE, TIME '09:32:05.123456')|
+------------------------------------------+
|                                  09:32:00|
+------------------------------------------+

scala> spark.sql("SELECT time_trunc('second', TIME'09:32:05.123456');").show()
+-----------------------------------+
|time_trunc(second, 09:32:05.123456)|
+-----------------------------------+
|                           09:32:05|
+-----------------------------------+

scala> spark.sql("SELECT time_trunc(concat('milli','second'), TIME'09:32:05.123456');").show()
+--------------------------------------------------+
|time_trunc(concat(milli, second), 09:32:05.123456)|
+--------------------------------------------------+
|                                      09:32:05.123|
+--------------------------------------------------+

scala> spark.sql("SELECT time_trunc('MICROSECOND', TIME'09:32:05.123456');").show()
+----------------------------------------+
|time_trunc(MICROSECOND, 09:32:05.123456)|
+----------------------------------------+
|                         09:32:05.123456|
+----------------------------------------+

scala> spark.sql("SELECT time_trunc('MICROSECOND', TIME'09:32:05.1234');").show()
+--------------------------------------+
|time_trunc(MICROSECOND, 09:32:05.1234)|
+--------------------------------------+
|                         09:32:05.1234|
+--------------------------------------+

# Invalid inputs
scala> spark.sql("SELECT time_trunc('MICROSECOND', '09:32:05.1234');").show()
org.apache.spark.sql.catalyst.ExtendedAnalysisException: [DATATYPE_MISMATCH.UNEXPECTED_INPUT_TYPE] Cannot resolve "time_trunc(MICROSECOND, 09:32:05.1234)" due to data type mismatch: The second parameter requires the "TIME" type, however "09:32:05.1234" has the type "STRING". SQLSTATE: 42K09; line 1 pos 7;
'Project [unresolvedalias(time_trunc(MICROSECOND, 09:32:05.1234))]
+- OneRowRelation

scala> spark.sql("SELECT time_trunc('MS', TIME'09:32:05.123456');").show()
+-------------------------------+
|time_trunc(MS, 09:32:05.123456)|
+-------------------------------+
|                           NULL|
+-------------------------------+

scala> spark.sql("SELECT time_trunc('MICROSECOND', TIME'29:32:05.123456');").show()
org.apache.spark.sql.catalyst.parser.ParseException:
[INVALID_TYPED_LITERAL] The value of the typed literal "TIME" is invalid: '29:32:05.123456'. SQLSTATE: 42604
== SQL (line 1, position 34) ==
...ELECT time_trunc('MICROSECOND', TIME'29:32:05.123456');
                                   ^^^^^^^^^^^^^^^^^^^^^
  at org.apache.spark.sql.errors.QueryParsingErrors$.cannotParseValueTypeError(QueryParsingErrors.scala:250)'

# unfoldable inputs
scala> val df = Seq(
     |   ("HOUR",       "09:32:05.123456"),
     |   ("MINUTE",     "10:20:15.123456"),
     |   ("second",     "11:59:59.999999"),
     |   ("MILLISECOND","00:00:00.123000"),
     |   ("MICROSECOND","00:00:00.123000"),
     |   ("MICROSECOND","00:00:00.123456"),
     |   ("MS","00:00:00.123456"),
     |   ("NULL","00:00:00.123456")
     | ).toDF("unitcol", "timecol")
val df: org.apache.spark.sql.DataFrame = [unitcol: string, timecol: string]

scala> val timeDf = df.selectExpr("unitcol", "CAST(timecol AS TIME(6)) as timeval")
val timeDf: org.apache.spark.sql.DataFrame = [unitcol: string, timeval: time(6)]

scala> timeDf.createOrReplaceTempView("tmp")

scala> spark.sql("""
     |   SELECT
     |     unitcol,
     |     timeval,
     |     time_trunc(unitcol, timeval) as truncated
     |   FROM tmp
     | """).show(false)
+-----------+---------------+---------------+
|unitcol    |timeval        |truncated      |
+-----------+---------------+---------------+
|HOUR       |09:32:05.123456|09:00:00       |
|MINUTE     |10:20:15.123456|10:20:00       |
|second     |11:59:59.999999|11:59:59       |
|MILLISECOND|00:00:00.123   |00:00:00.123   |
|MICROSECOND|00:00:00.123   |00:00:00.123   |
|MICROSECOND|00:00:00.123456|00:00:00.123456|
|MS         |00:00:00.123456|NULL           |
|NULL       |00:00:00.123456|NULL           |
+-----------+---------------+---------------+
```


### Was this patch authored or co-authored using generative AI tooling?
No.

the-sakthi committed:
[SPARK-51554][SQL] Add the time_trunc() function for TIME datatype

the-sakthi commented:
@MaxGekk While I add the UTs for this, would appreciate any feedbacks from you on this, meanwhile!

the-sakthi committed:
Extending runtimerepalaceble

the-sakthi committed:
Fix expression schema

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:
```diff
@@ -532,6 +534,61 @@ object DateTimeUtils extends SparkDateTimeUtils {
+    if (level < MIN_LEVEL_OF_TIME_TRUNC) {  // unknown unit
+      -1L
+    } else {
+      val out = truncTime(micros, level)         // may still return -1L on error
```
What's the reason of splitting the function? Is `truncTime` used somewhere else?

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/timeExpressions.scala:
```diff
@@ -548,3 +548,77 @@ case class MakeTime(
+  override def inputTypes: Seq[AbstractDataType] =
+    Seq(
+      StringType,
+      TypeCollection(TimeType.MIN_PRECISION to TimeType.MAX_PRECISION map TimeType: _*)
```
Could you double check that it won't produce warnings.

the-sakthi committed:
revision comments 1

- rebased with latest changes
- change in internal representation of time data type
